File: windows/src/Zharp.App/Themes.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Media;
using Windows.UI;
using Zharp.Core.Terminal;

namespace Zharp.App;

/// <summary>One selectable color theme: chrome colors + terminal palette.</summary>
public sealed record ThemeSpec(
    string Id,
    string Name,
    bool IsDark,
    uint ChromeBackground,
    uint IconColor,
    Func<Palette> CreatePalette);

/// <summary>
/// Theme registry. Chrome brushes live in App.xaml's two ThemeDictionaries
/// (Dark/Light buckets provide the alpha overlays); applying a theme picks the
/// bucket via RequestedTheme and MUTATES the shared brushes' colors in place -
/// every {ThemeResource} reference updates live.
/// </summary>
public static class Themes
{
    public static readonly ThemeSpec[] All =
    [
        new("cream", "Cream", false, 0xEDE6D8, 0x2A2C33, Palette.Cream),
        new("paper", "Paper", false, 0xF6F5F1, 0x24292F, Palette.Paper),
        new("rose", "Rosé", false, 0xFAF4ED, 0x575279, Palette.Rose),
        new("dark", "Dark", true, 0x282828, 0xFFFFFF, Palette.Campbell),
        new("navy", "Navy", true, 0x151E32, 0xD5DEF2, Palette.Navy),
        new("tokyo", "Tokyo", true, 0x1A1B26, 0xC0CAF5, Palette.Tokyo),
        new("dracula", "Dracula", true, 0x282A36, 0xF8F8F2, Palette.Dracula),
        new("catppuccin", "Catppuccin", true, 0x1E1E2E, 0xCDD6F4, Palette.Catppuccin),
        new("gruvbox", "Gruvbox", true, 0x1D2021, 0xEBDBB2, Palette.Gruvbox),
    ];

    public static ThemeSpec Get(string? id) =>
        All.FirstOrDefault(t => string.Equals(t.Id, id, StringComparison.OrdinalIgnoreCase)) ?? All[0];

    /// <summary>
    /// Writes the theme's colors into the active bucket's shared brushes.
    /// <paramref name="backgroundOpacity"/> (0.5-1.0) sets the chrome wash alpha -
    /// lower values let the Mica/Acrylic backdrop show through.
    /// </summary>
    public static void ApplyToResources(ThemeSpec theme, double backgroundOpacity)
    {
        var dictionaries = Application.Current.Resources.ThemeDictionaries;
        if (!dictionaries.TryGetValue(theme.IsDark ? "Dark" : "Light", out object? bucketObj) ||
            bucketObj is not ResourceDictionary bucket)
        {
            return;
        }

        byte washAlpha = (byte)Math.Clamp(backgroundOpacity * 255, 100, 255);
        SetBrush(bucket, "ChromeWashBrush", theme.ChromeBackground, washAlpha);
        SetBrush(bucket, "BarIconBrush", theme.IconColor, 0x9E);
        SetBrush(bucket, "SubtleIconBrush", theme.IconColor, 0x70);
        // Floating panels are opaque and slightly lifted from the chrome color,
        // so they stay readable over terminal text.
        SetBrush(bucket, "FloatingPanelBrush",
            Lift(theme.ChromeBackground, theme.IsDark ? 0.06 : 0.45), 0xFF);
    }

    /// <summary>Blends an RGB color toward white by the given fraction.</summary>
    private static uint Lift(uint rgb, double toWhite)
    {
        uint Channel(int shift)
        {
            uint c = (rgb >> shift) & 0xFF;
            return (uint)Math.Clamp(c + (255 - c) * toWhite, 0, 255);
        }
        return (Channel(16) << 16) | (Channel(8) << 8) | Channel(0);
    }

    private static void SetBrush(ResourceDictionary bucket, string key, uint rgb, byte alpha)
    {
        if (bucket.TryGetValue(key, out object? value) && value is SolidColorBrush brush)
        {
            brush.Color = Color.FromArgb(alpha, (byte)(rgb >> 16), (byte)(rgb >> 8), (byte)rgb);
        }
    }
}

File: macos/Sources/ZharpApp/Themes.swift
import AppKit
import ZharpCore

/// One selectable color theme: chrome colors + terminal palette.
struct ThemeSpec {
    let id: String
    let name: String
    let isDark: Bool
    let chromeBackground: UInt32
    let iconColor: UInt32
    let createPalette: () -> Palette
}

/// Theme registry. Where the Windows build mutates the brushes in App.xaml's
/// theme dictionaries, here `ChromeColors` is rebuilt from the active spec and
/// every view reads its colors from `Chrome.current` on `applyTheme`.
enum Themes {
    static let all: [ThemeSpec] = [
        ThemeSpec(id: "cream", name: "Cream", isDark: false,
                  chromeBackground: 0xEDE6D8, iconColor: 0x2A2C33, createPalette: Palette.cream),
        ThemeSpec(id: "paper", name: "Paper", isDark: false,
                  chromeBackground: 0xF6F5F1, iconColor: 0x24292F, createPalette: Palette.paper),
        ThemeSpec(id: "rose", name: "Rosé", isDark: false,
                  chromeBackground: 0xFAF4ED, iconColor: 0x575279, createPalette: Palette.rose),
        ThemeSpec(id: "dark", name: "Dark", isDark: true,
                  chromeBackground: 0x282828, iconColor: 0xFFFFFF, createPalette: Palette.campbell),
        ThemeSpec(id: "navy", name: "Navy", isDark: true,
                  chromeBackground: 0x151E32, iconColor: 0xD5DEF2, createPalette: Palette.navy),
        ThemeSpec(id: "tokyo", name: "Tokyo", isDark: true,
                  chromeBackground: 0x1A1B26, iconColor: 0xC0CAF5, createPalette: Palette.tokyo),
        ThemeSpec(id: "dracula", name: "Dracula", isDark: true,
                  chromeBackground: 0x282A36, iconColor: 0xF8F8F2, createPalette: Palette.dracula),
        ThemeSpec(id: "catppuccin", name: "Catppuccin", isDark: true,
                  chromeBackground: 0x1E1E2E, iconColor: 0xCDD6F4, createPalette: Palette.catppuccin),
        ThemeSpec(id: "gruvbox", name: "Gruvbox", isDark: true,
                  chromeBackground: 0x1D2021, iconColor: 0xEBDBB2, createPalette: Palette.gruvbox),
    ]

    static func get(_ id: String?) -> ThemeSpec {
        all.first { $0.id.caseInsensitiveCompare(id ?? "") == .orderedSame } ?? all[0]
    }
}

/// The chrome brush set - the direct counterpart of App.xaml's
/// ChromeWash / PanelOverlay / Hairline / IconChip / BarIcon / SubtleIcon /
/// SectionDivider / FloatingPanel brushes.
struct ChromeColors {
    var isDark: Bool
    var chromeWash: NSColor
    var panelOverlay: NSColor
    var hairline: NSColor
    var iconChip: NSColor
    var barIcon: NSColor
    var subtleIcon: NSColor
    var sectionDivider: NSColor
    var floatingPanel: NSColor
    var text: NSColor
    var accent: NSColor
    /// Hover / selection backplates for list rows.
    var rowHover: NSColor
    var rowSelected: NSColor
    /// Brand wordmark color: cream on dark surfaces, ink on light ones.
    var brandForeground: NSColor
    /// The brand mark variant that reads on this theme's surface.
    var brandLogoName: String
    /// Update-available badge in the title bar.
    var updateBadgeBackground: NSColor
    var updateBadgeForeground: NSColor

    static func make(_ theme: ThemeSpec, backgroundOpacity: Double) -> ChromeColors {
        let washAlpha = min(max(backgroundOpacity, 100.0 / 255.0), 1.0)
        let dark = theme.isDark
        let palette = theme.createPalette()
        return ChromeColors(
            isDark: dark,
            chromeWash: rgb(theme.chromeBackground, alpha: washAlpha),
            panelOverlay: dark ? white(0, 0x28) : white(0, 0x12),
            hairline: dark ? white(0xFF, 0x16) : white(0, 0x21),
            iconChip: dark ? white(0xFF, 0x1E) : white(0, 0x14),
            barIcon: rgb(theme.iconColor, alpha: Double(0x9E) / 255.0),
            subtleIcon: rgb(theme.iconColor, alpha: Double(0x70) / 255.0),
            sectionDivider: dark ? white(0xFF, 0x1A) : white(0, 0x20),
            // Floating panels are opaque and slightly lifted from the chrome
            // color, so they stay readable over terminal text.
            floatingPanel: rgb(lift(theme.chromeBackground, toWhite: dark ? 0.06 : 0.45), alpha: 1.0),
            text: rgb(palette.defaultForeground, alpha: 1.0),
            accent: rgb(palette.cursorColor, alpha: 1.0),
            rowHover: dark ? white(0xFF, 0x14) : white(0, 0x0E),
            rowSelected: dark ? white(0xFF, 0x24) : white(0, 0x1A),
            brandForeground: rgb(dark ? 0xF0EFEC : 0x1A1A19, alpha: 1),
            brandLogoName: dark ? "logo-cream" : "logo-ink",
            // A subtle tinted pill: red enough to notice, quiet enough to live
            // in the title bar without reading as an error.
            updateBadgeBackground: dark
                ? rgb(0xE0524D, alpha: Double(0x26) / 255.0)
                : rgb(0xC4433B, alpha: Double(0x1C) / 255.0),
            updateBadgeForeground: rgb(dark ? 0xEE9C95 : 0xA8423B, alpha: 1)
        )
    }

    /// Blends an RGB color toward white by the given fraction.
    private static func lift(_ rgbValue: UInt32, toWhite: Double) -> UInt32 {
        func channel(_ shift: UInt32) -> UInt32 {
            let c = Double((rgbValue >> shift) & 0xFF)
            return UInt32(min(max(c + (255 - c) * toWhite, 0), 255))
        }
        return (channel(16) << 16) | (channel(8) << 8) | channel(0)
    }

    static func rgb(_ value: UInt32, alpha: Double) -> NSColor {
        NSColor(srgbRed: CGFloat((value >> 16) & 0xFF) / 255.0,
                green: CGFloat((value >> 8) & 0xFF) / 255.0,
                blue: CGFloat(value & 0xFF) / 255.0,
                alpha: CGFloat(alpha))
    }

    private static func white(_ level: UInt32, _ alpha: UInt32) -> NSColor {
        rgb(level == 0 ? 0x000000 : 0xFFFFFF, alpha: Double(alpha) / 255.0)
    }
}

/// Ambient chrome palette, read by every view when it draws.
enum Chrome {
    static var current = ChromeColors.make(Themes.all[0], backgroundOpacity: 0.94)
}
